### PR TITLE
[#PAB-332] DB model changes and R3 transformation improvements

### PR DESCRIPTION
**Pull request by nsnape**
https://trello.com/c/97RfNJ9h

### Change description
Improvements to R3 transformation pipeline:

- Removed redundant code
- vectorise some slow iterative methods
- Add TODO comments for parts to review (how and when to consider an input row "empty" and whether to discard it upon ingest)
- Changes to model - nullability on value fields rather than defaulting to 0, to persist the actual value users entered (or didn't)

- [x] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### CODE REVIEW

File: core/extraction/towns_fund.py [hunk 1]
```diff
@@ -562,7 +569,7 @@ def extract_risks(df_risk: pd.DataFrame, project_lookup: dict, programme_id: str
         "Proximity",
         "RiskOwnerRole",
     ]
-
+    # TODO: check this drop behaviour. Should we only drop if whole row is empty / null?
```
Comment (cyrusdobbs) [inline core/extraction/towns_fund.py:572]: Just checking these TODOs are to think about later rather than as part of this PR?

Comment (cyrusdobbs) [reply core/extraction/towns_fund.py:572]: Does it need to be captured in a ticket?

**Review by cyrusdobbs — CHANGES_REQUESTED** [review]
These changes look fine just that ingest is broken.. and making sure those TODOs are ok to be left as TODOs

File: core/db/entities.py [hunk 2]
```diff
@@ -398,10 +398,9 @@ class PrivateInvestment(BaseModel):
     project_id: Mapped[GUID] = sqla.orm.mapped_column(sqla.ForeignKey("project_dim.id"), nullable=False)
 
     total_project_value = sqla.Column(sqla.Float(), nullable=False)
-    # TODO: Review whether these should be defaulted to 0, or actually nullable?
-    townsfund_funding = sqla.Column(sqla.Float(), nullable=False, default=0.0)
-    private_sector_funding_required = sqla.Column(sqla.Float(), nullable=False, default=0.0)
-    private_sector_funding_secured = sqla.Column(sqla.Float(), nullable=False, default=0.0)
+    townsfund_funding = sqla.Column(sqla.Float(), nullable=False)
+    private_sector_funding_required = sqla.Column(sqla.Float(), nullable=True)
+    private_sector_funding_secured = sqla.Column(sqla.Float(), nullable=True)
```
[the remark this answers is not in the document]
Comment (cyrusdobbs) [reply core/db/entities.py:403]: Or a separate fix branch before this goes in?

File: core/extraction/towns_fund.py (hunk 1 above)
Comment (nsnape) [reply core/extraction/towns_fund.py:572]: Yeah, i'll leave them in for now, they are on me to check with Maeve and Enda

Comment (nsnape) [reply core/extraction/towns_fund.py:572]: I'll add tickets for them at some point

File: core/db/entities.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nsnape) [reply core/db/entities.py:403]: done

**Review by cyrusdobbs — APPROVED** [review]
LGTM

**Pull request merged**